From 5407fba61f6d0d673b87cfa42e77deec70438b7c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Tiziano=20M=C3=BCller?= <tiziano.mueller@hpe.com>
Date: Tue, 24 Mar 2026 16:28:18 +0100
Subject: [PATCH] build: fix parallel build issues
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

These show up when building on ARM for both 4k and 64k pagesize kernels.

Building the common object twice might be a bit wasteful, but greatly
simplifies the dependencies.

Signed-off-by: Tiziano Mueller <tiziano.mueller@hpe.com>
---
diff --git a/Makefile b/Makefile
--- a/Makefile
+++ b/Makefile
@@ -73,7 +73,5 @@ nroff:
 else
 
-.NOTPARALLEL: tests/cxi/single_client/ tests/cxi/multi_client/
-
 obj-y += kfi/
 obj-y += prov/cxi/
 #obj-y += prov/ibverbs/
diff --git a/tests/cxi/multi_client/Makefile b/tests/cxi/multi_client/Makefile
--- a/tests/cxi/multi_client/Makefile
+++ b/tests/cxi/multi_client/Makefile
@@ -1,11 +1,15 @@
 TOPDIR := $(src)/../../..
-TEST_COMMON_DIR := $(TOPDIR)/tests/cxi/common
-TEST_COMMON_LIB := ../common/test_common.o
 
 obj-m := multi-nic-msg.o
 
-multi-nic-msg-objs := multi_nic_msg.o $(TEST_COMMON_LIB)
+multi-nic-msg-objs := multi_nic_msg.o test_common.o
 
 ccflags-y += -I$(TOPDIR)/include -I$(TOPDIR)/../cxi-driver/include \
 	-I$(TOPDIR)/../cassini-headers/install/include \
-	-I$(TEST_COMMON_DIR)
+	-I$(TOPDIR)/tests/cxi/common
+
+# Compile test_common.c from the common directory into this build directory
+# so that parallel builds of single_client/ and multi_client/ do not race
+# on a shared output file.
+$(obj)/test_common.o: $(src)/../common/test_common.c FORCE
+	$(call if_changed_rule,cc_o_c)
diff --git a/tests/cxi/single_client/Makefile b/tests/cxi/single_client/Makefile
--- a/tests/cxi/single_client/Makefile
+++ b/tests/cxi/single_client/Makefile
@@ -1,6 +1,4 @@
 TOPDIR := $(src)/../../..
-TEST_COMMON_DIR := $(TOPDIR)/tests/cxi/common
-TEST_COMMON_LIB := ../common/test_common.o
 
 obj-m := \
 	test_tagged_ops.o \
@@ -48,14 +46,20 @@ obj-m := \
 	test_rma_md_cache.o
 
-test-msg-ops-thread-objs := test_msg_ops_thread.o $(TEST_COMMON_LIB)
-test-rma-ops-thread-objs := test_rma_ops_thread.o $(TEST_COMMON_LIB)
-test-mr-alloc-async-objs := test_mr_alloc_async.o $(TEST_COMMON_LIB)
-test-mr-alloc-sync-objs := test_mr_alloc_sync.o $(TEST_COMMON_LIB)
-test-mr-alloc-async-thread-objs := test_mr_alloc_async_thread.o $(TEST_COMMON_LIB)
-test-rx-ctx-opts-objs := test_rx_ctx_opts.o $(TEST_COMMON_LIB)
-test-tx-ctx-opts-objs := test_tx_ctx_opts.o $(TEST_COMMON_LIB)
-test-sep-rx-tx-ctx-objs := test_sep_rx_tx_ctx.o $(TEST_COMMON_LIB)
+test-msg-ops-thread-objs := test_msg_ops_thread.o test_common.o
+test-rma-ops-thread-objs := test_rma_ops_thread.o test_common.o
+test-mr-alloc-async-objs := test_mr_alloc_async.o test_common.o
+test-mr-alloc-sync-objs := test_mr_alloc_sync.o test_common.o
+test-mr-alloc-async-thread-objs := test_mr_alloc_async_thread.o test_common.o
+test-rx-ctx-opts-objs := test_rx_ctx_opts.o test_common.o
+test-tx-ctx-opts-objs := test_tx_ctx_opts.o test_common.o
+test-sep-rx-tx-ctx-objs := test_sep_rx_tx_ctx.o test_common.o
 
 ccflags-y += -I$(TOPDIR)/include -I$(TOPDIR)/../cxi-driver/include \
 	-I$(TOPDIR)/../cassini-headers/install/include \
-	-I$(TEST_COMMON_DIR)
+	-I$(TOPDIR)/tests/cxi/common
+
+# Compile test_common.c from the common directory into this build directory
+# so that parallel builds of single_client/ and multi_client/ do not race
+# on a shared output file.
+$(obj)/test_common.o: $(src)/../common/test_common.c FORCE
+	$(call if_changed_rule,cc_o_c)
